### PR TITLE
WT-9241 Process the call log for timestamp_transaction_uint()

### PR DESCRIPTION
**Pull request by Sean04**
I've added the processing logic in the call log manager to invoke the timestamp_transaction_uint() API in the session. At the moment the set methods for the timestamps are placeholders and the validation logic for the timestamps will be handled in future tickets. 

I've also removed the quotes in the generate call log for timestamp_transaction_uint() for the timestamp input argument so that it can be interpreted as a uint64_t instead of a string.

### CODE REVIEW

File: test/simulator/timestamp/call_log_manager/call_log_manager.cpp [hunk 1]
```diff
@@ -226,6 +227,32 @@ call_log_manager::call_log_set_timestamp(const json &call_log_entry)
...
+void
+call_log_manager::call_log_timestamp_transaction_uint(const json &call_log_entry)
+{
+    /* Convert the transaction type char * to a string object. */
+    const std::string wt_ts_txn_type = call_log_entry["input"]["wt_ts_txp_type"].get<std::string>();
+
+    /*
+     * There are no timestamps to be set if the timestamp type is not specified.
+     */
+    if (wt_ts_txn_type == "unkown")
```
Comment (bvpvamsikrishna) [inline test/simulator/timestamp/call_log_manager/call_log_manager.cpp:239]: I think it is better we have an error msg here, to indicate that timestamp hasn't been set.

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm, if we add the error msg.

File: test/simulator/timestamp/call_log_manager/call_log_manager.cpp (hunk 1 above)
Comment (luke-pearson) [inline test/simulator/timestamp/call_log_manager/call_log_manager.cpp:239]: unkown -> unknown.

Comment (Sean04) [reply test/simulator/timestamp/call_log_manager/call_log_manager.cpp:239]: From my understanding, wiredtiger just silently ignores this case, so there wouldn't be any particular message to compare the output to. Should I still add a message regardless?

File: test/simulator/timestamp/call_log_manager/call_log_manager.cpp [hunk 2]
```diff
@@ -226,6 +227,32 @@ call_log_manager::call_log_set_timestamp(const json &call_log_entry)
...
 }
 
+void
+call_log_manager::call_log_timestamp_transaction_uint(const json &call_log_entry)
+{
+    /* Convert the transaction type char * to a string object. */
+    const std::string wt_ts_txn_type = call_log_entry["input"]["wt_ts_txp_type"].get<std::string>();
+
+    /*
+     * There are no timestamps to be set if the timestamp type is not specified.
```
Comment (etienneptl) [inline test/simulator/timestamp/call_log_manager/call_log_manager.cpp:237]: Can this be a one-line comment?

File: test/simulator/timestamp/src/session_simulator.cpp [hunk 3]
```diff
@@ -59,3 +60,48 @@ session_simulator::commit_transaction()
...
+
+    if (ts_type == "commit") {
+        set_commit_timestamp(ts);
+    } else if (ts_type == "durable") {
+        set_durable_timestamp(ts);
+    } else if (ts_type == "prepare") {
+        set_prepare_timestamp(ts);
+    } else if (ts_type == "read") {
+        set_read_timestamp(ts);
+    }
```
Comment (etienneptl) [inline test/simulator/timestamp/src/session_simulator.cpp:104]: We should probably throw an error in an `else` statement.

File: test/simulator/timestamp/src/include/session_simulator.h [hunk 4]
```diff
@@ -36,6 +38,11 @@ class session_simulator {
     void begin_transaction();
     void rollback_transaction();
     void commit_transaction();
+    int timestamp_transaction_uint(const std::string &, const uint64_t &);
+    void set_commit_timestamp(const uint64_t &);
```
Comment (etienneptl) [inline test/simulator/timestamp/src/include/session_simulator.h:42]: As discussed, we can simply pass the argument as non-const and by value. By definition, a setter should not do anything than use the value and assign it to a private member.

File: test/simulator/timestamp/src/session_simulator.cpp (hunk 3 above)
Comment (Sean04) [reply test/simulator/timestamp/src/session_simulator.cpp:104]: Yeah, that's a good idea.

Comment (Sean04) [reply test/simulator/timestamp/src/session_simulator.cpp:104]: That relates to Vamsi's comment about an error earlier. WiredTiger just silently ignores if it doesn't have a correct timestamp type. @bvpvamsikrishna Should we add an error in this case?

File: test/simulator/timestamp/src/include/session_simulator.h (hunk 4 above)
Comment (Sean04) [reply test/simulator/timestamp/src/include/session_simulator.h:42]: What about in `timestamp_transaction_uint`?

File: test/simulator/timestamp/src/session_simulator.cpp [hunk 5]
```diff
@@ -59,3 +60,48 @@ session_simulator::commit_transaction()
...
+session_simulator::set_read_timestamp(const uint64_t &ts)
+{
+    _read_ts = ts;
+}
+
+int
+session_simulator::timestamp_transaction_uint(const std::string &ts_type, const uint64_t &ts)
+{
+    /* Zero timestamp is not permitted. */
+    if (ts == 0) {
```
Comment (etienneptl) [inline test/simulator/timestamp/src/session_simulator.cpp:92]: Could this be an assertion instead? It's fine as it is but sometimes we do asserts, sometimes `if` and a message.

File: test/simulator/timestamp/src/include/session_simulator.h (hunk 4 above)
Comment (etienneptl) [reply test/simulator/timestamp/src/include/session_simulator.h:42]: `session_simulator::timestamp_transaction_uint(const std::string &ts_type, uint64_t ts)`

File: test/simulator/timestamp/src/session_simulator.cpp (hunk 5 above)
Comment (Sean04) [reply test/simulator/timestamp/src/session_simulator.cpp:92]: That's true, I think an assert would be more consistent.

Comment (Sean04) [reply test/simulator/timestamp/src/session_simulator.cpp:92]: Actually, sorry, we want to match up the return value from the simulator with the call log entry generated from wiredtiger so in the case of having a 0 timestamp we can check that `EINVAL` is returned.

File: test/simulator/timestamp/src/include/session_simulator.h (hunk 4 above)
Comment (Sean04) [reply test/simulator/timestamp/src/include/session_simulator.h]: Fixed.

File: test/simulator/timestamp/call_log_manager/call_log_manager.cpp (hunk 2 above)
Comment (Sean04) [reply test/simulator/timestamp/call_log_manager/call_log_manager.cpp]: Fixed.

File: test/simulator/timestamp/call_log_manager/call_log_manager.cpp (hunk 1 above)
Comment (Sean04) [reply test/simulator/timestamp/call_log_manager/call_log_manager.cpp]: Fixed.

**Review by etienneptl — APPROVED** [review]
LGTM after the comment about `ts_type` is fixed. Good job!

**Review by raviprakashgiri29 — APPROVED** [review]
The overall changes LGTM

File: test/simulator/timestamp/src/session_simulator.cpp (hunk 3 above)
Comment (bvpvamsikrishna) [reply test/simulator/timestamp/src/session_simulator.cpp:104]: Yes, please add an error.

Comment (Sean04) [reply test/simulator/timestamp/src/session_simulator.cpp:107]: Added.

**Pull request merged**